kedro/runner/runner.py

```diff
@@ -99,9 +101,13 @@ def run(
             )
 
         # Identify MemoryDataset in the catalog
+        catalog_datasets = (
```

Comment:
That's exactly the kind of thing an abstract DataCatalog (or a Protocol to this end) will prevent. Having code to check what type of instance we are working with is a bit of an anti-pattern in most cases (bar some exceptional cases emulating pattern matching). Could we somehow create a minimal interface which is used within Kedro, give it a name and program against it?